beacon_node/beacon_chain/src/attester_cache.rs

```diff
@@ -0,0 +1,380 @@
+    /// the state identified by the cache `key`.
+    pub fn get<T: EthSpec>(
+        &self,
+        key: &AttesterCacheKey,
+        slot: Slot,
+        committee_index: CommitteeIndex,
+        spec: &ChainSpec,
+    ) -> Result<Option<(JustifiedCheckpoint, CommitteeLength)>, Error> {
+        self.cache
+            .read()
+            .get(key)
+            .map(|cache_item| cache_item.get::<T>(slot, committee_index, spec))
+            .transpose()
+    }
+
+    /// Cache the `state.current_epoch()` values if they are not already present in the state.
+    pub fn maybe_cache_state<T: EthSpec>(
+        &self,
+        state: &BeaconState<T>,
+        latest_block_root: Hash256,
+        spec: &ChainSpec,
+    ) -> Result<(), Error> {
+        let key = AttesterCacheKey::new(state.current_epoch(), state, latest_block_root)?;
+        let mut cache = self.cache.write();
+        if !cache.contains_key(&key) {
+            let cache_item = AttesterCacheValue::new(state, spec)?;
+            Self::insert_respecting_max_len(&mut cache, key, cache_item);
+        }
+        Ok(())
+    }
+
+    /// Read the state identified by `state_root` from the database, advance it to the required
+    /// slot, use it to prime the cache and return the values for the provided `slot` and
+    /// `committee_index`.
+    ///
+    /// ## Notes
+    ///
+    /// This function takes a write-lock on the internal cache. Prefer attempting a `Self::get` call
+    /// before running this function as `Self::get` only takes a read-lock and is therefore less
+    /// likely to create head contention.
+    pub fn load_and_cache_state<T: BeaconChainTypes>(
+        &self,
+        state_root: Hash256,
+        key: AttesterCacheKey,
+        slot: Slot,
+        committee_index: CommitteeIndex,
+        chain: &BeaconChain<T>,
+    ) -> Result<(JustifiedCheckpoint, CommitteeLength), Error> {
+        let spec = &chain.spec;
+        let slots_per_epoch = T::EthSpec::slots_per_epoch();
+        let epoch = slot.epoch(slots_per_epoch);
+
+        // Take a write-lock on the cache before starting the state read.
+        //
+        // Whilst holding the write-lock during the state read will create contention, it prevents
+        // the scenario where multiple requests from separate threads cause duplicate state reads.
+        let mut cache = self.cache.write();
+
+        // Try the cache to see if someone has already primed it between the time the function was
+        // called and when the cache write-lock was obtained. This avoids performing duplicate state
+        // reads.
+        if let Some(value) = cache
+            .get(&key)
+            .map(|cache_item| cache_item.get::<T::EthSpec>(slot, committee_index, spec))
+            .transpose()?
+        {
+            return Ok(value);
+        }
+
+        let mut state: BeaconState<T::EthSpec> = chain
+            .get_state(&state_root, None)?
```

Comment:
I think we should (continue to) use `chain.state_at_slot` here, to avoid re-reading the state from the DB in the case where the state advance timer fails to prime the cache. One case where this will happen is if there are two consecutive skipped slots either side of the epoch boundary (slot 31 and slot 0 mod 32).

Reply:
`BeaconChain::state_at_slot` is racey since it takes another read-lock on the head without checking to see if the head is still relevant to the desired `state_root`.

I'd avoided cloning the head for this reason, but I just thought of a fairly tidy way to do it in 8cb488efe5cd9fdd628bc6b7edb0cffd2ed980b7. Let me know what you think. 

> (continue to)

I'm not sure what you mean by this. With the present code we were never using `state_at_slot`.

Reply:
Actually, I need to this about this a bit more. I'll ping you when I'm happy with it.

Reply:
Alright, I spent some more time with this and I ended up and in a similar position as before. I'll do my best to explain:

There's a property I want to maintain with this cache, let's call it "no double loads". Lets say a BN has enough validators to span all committee indices and they all hit `validator/attestation_data` at exactly the same time. If there is a cache miss, I want to avoid the scenario where we load/clone a `BeaconState` for each of those committees (i.e., 64x). I want the only *first* request to trigger a state load and to block all the other requests on that. You probably saw comments about this in `AttesterCache::load_and_cache_state`. 

Another property I want to maintain is "no lock interleaving". I.e., don't hold the `chain.canonical_head` and `chain.attester_cache.cache` locks at the same time.

Unfortunately, I can't see a simple and efficient solution where we maintain these properties *and* avoid the DB read by cloning the head.

I think we could get away with interleaving with the present code-base, but I'm concerned about future regressions. I'm interested in your thoughts.

> One case where this will happen is if there are two consecutive skipped slots either side of the epoch boundary (slot 31 and slot 0 mod 32).

My current thinking is that if we start to get burned by this then we should make the state advance timer keep advancing the state for each skipped slot (rather than just advancing once). I can't recall why we don't already do this, but it seems like a good idea to me :thinking:

Reply:
> My current thinking is that if we start to get burned by this then we should make the state advance timer keep advancing the state for each skipped slot (rather than just advancing once). I can't recall why we don't already do this, but it seems like a good idea to me thinking

I remembered. It's because we need to store intermediate states.